src/Dictionaries/TrieDictionary.cpp

```diff
@@ -31,20 +24,175 @@ namespace ErrorCodes
     extern const int TYPE_MISMATCH;
     extern const int BAD_ARGUMENTS;
     extern const int DICTIONARY_IS_EMPTY;
-    extern const int NOT_IMPLEMENTED;
+}
+
+namespace
+{
+    /// Intermediate structure that are used in loading procedure
+    struct IPRecord
+    {
+        Poco::Net::IPAddress addr;
+        UInt8 prefix;
+        size_t row;
+        bool isv6;
+
+        IPRecord(const Poco::Net::IPAddress & addr_, UInt8 prefix_, size_t row_)
+            : addr(addr_)
+            , prefix(prefix_)
+            , row(row_)
+            , isv6(addr.family() == Poco::Net::IPAddress::IPv6)
+        {
+        }
+
+        const uint8_t * asIPv6Binary(uint8_t * buf) const
+        {
+            if (isv6)
+                return reinterpret_cast<const uint8_t *>(addr.addr());
+            memset(buf, 0, 10);
+            buf[10] = '\xFF';
+            buf[11] = '\xFF';
+            memcpy(&buf[12], addr.addr(), 4);
+
+            return buf;
+        }
+
+        inline UInt8 prefixIPv6() const
+        {
+            return isv6 ? prefix : prefix + 96;
+        }
+    };
+
+    struct IPv4Subnet
+    {
+        UInt32 addr;
+        UInt8 prefix;
+    };
+
+    struct IPv6Subnet
+    {
+        const uint8_t * addr;
+        UInt8 prefix;
+    };
 }
 
 static void validateKeyTypes(const DataTypes & key_types)
 {
-    if (key_types.size() != 1)
-        throw Exception{"Expected a single IP address", ErrorCodes::TYPE_MISMATCH};
+    if (key_types.empty() || key_types.size() > 2)
+        throw Exception{"Expected a single IP address or IP with mask", ErrorCodes::TYPE_MISMATCH};
 
     const auto & actual_type = key_types[0]->getName();
-
     if (actual_type != "UInt32" && actual_type != "FixedString(16)")
         throw Exception{"Key does not match, expected either UInt32 or FixedString(16)", ErrorCodes::TYPE_MISMATCH};
+
+    if (key_types.size() > 1)
+    {
+        const auto * mask_col_type = typeid_cast<const DataTypeUInt8 *>(key_types[1].get());
+        if (mask_col_type == nullptr)
+            throw Exception{"Mask do not match, expected UInt8", ErrorCodes::TYPE_MISMATCH};
+    }
+}
+
+template <typename T, typename Comp>
+size_t sort_and_unique(std::vector<T> & vec, Comp comp)
+{
+    std::sort(vec.begin(), vec.end(),
+              [&](const auto & a, const auto & b) { return comp(a, b) < 0; });
+
+    auto new_end = std::unique(vec.begin(), vec.end(),
+                               [&](const auto & a, const auto & b) { return comp(a, b) == 0; });
+    if (new_end != vec.end())
+    {
+        vec.erase(new_end, vec.end());
+        return std::distance(new_end, vec.end());
```

Comment:
Looks like `new_end` is equals to `vec.end()` after `vec.erase`, so it always return zero

Reply:
Also dosc say that erase `Invalidates iterators and references at or after the point of the erase`, so `new_end` may be invalid ...